File: Practice/QuickSort.py
def quick_sort(arr, left, right):
    if left < right:
        partiton_pos = partiton(arr, left, right)
        quick_sort(arr, left, partiton_pos - 1)
        quick_sort(arr, partiton_pos + 1, right)

def partiton(arr, left, right):
    i = left
    j = right - 1
    pivot = arr[right]

    while i < j:
        while i < right and arr[i] < pivot:
            i = i + 1
        while j > left and arr[j] >= pivot:
            j = j - 1
        if i < j:
            arr[i], arr[j] = arr[j], arr[i]

    if arr[i] > pivot:
        arr[i], arr[right] = arr[right], arr[i]

    return i
        
print("Enter number of values:")
n = int(input())

print("Enter list to sort:")
lst = []
for i in range(0,n):
    ele = int(input())

    lst.append(ele)
print(lst)

quick_sort(lst, 0, len(lst) - 1)

print("Sorted array:")
print(lst)

File: Lab/Practice2/QuickSort.cpp
#include <stdio.h>
#include <iostream>
using namespace std;

// function to swap elements
void swap(int *a, int *b) {
  int t = *a;
  *a = *b;
  *b = t;
}

// function to print the array
void printArray(int array[], int size) {
  int i;
  for (i = 0; i < size; i++)
    cout << array[i] << " ";
  cout << endl;
}

// function to rearrange array (find the partition point)
int partition(int array[], int low, int high) {
    
  // select the rightmost element as pivot
  int pivot = array[high];
  
  // pointer for greater element
  int i = (low - 1);

  for (int j = low; j < high; j++) {
    if (array[j] <= pivot) {
      
      i++;
      
      // swap element at i with element at j
      swap(&array[i], &array[j]);
    }
  }
  
  // swap pivot with the greater element at i
  swap(&array[i + 1], &array[high]);
  
  // return the partition point
  return (i + 1);
}

void quickSort(int array[], int low, int high) {
  if (low < high) {
    int pi = partition(array, low, high);

    // recursive call on the left of pivot
    quickSort(array, low, pi - 1);

    // recursive call on the right of pivot
    quickSort(array, pi + 1, high);
  }
}

int main(){
    int n, data[100];

    cout<<"Enter size";
    cin >> n;

    cout<<"Enter the array:";

    for(int i = 0; i < n; i++){
        cin>>data[i];
    }

    cout << "Unsorted Array: \n";
    printArray(data, n);
  
    // perform quicksort on data
    quickSort(data, 0, n - 1);
  
    cout << "Sorted array in ascending order: \n";
    printArray(data, n);
}